File: SDK/WebClient/src/APIDefinitions/IAccountApi.ts
import { ApiDefinition, methodStub } from '@remote-ioc/runtime';
import { AccountInfo, TypedEventEmitter } from '../Types';

@ApiDefinition('account-api')
export class IAccountApi extends TypedEventEmitter<{ updated: [string] }> {
  public async getAccounts(): Promise<AccountInfo[]> {
    methodStub(this);
  }

  public async getAccountDump(accountId: string): Promise<{}> {
    methodStub(this, accountId);
  }

  public async getArtifacts(accountId: string): Promise<unknown[]> {
    methodStub(this, accountId);
  }

  public async getArtifactById(accountId: string, artifactId: number): Promise<unknown> {
    methodStub(this, accountId, artifactId);
  }

  public async getHeroes(accountId: string, snapshot?: boolean): Promise<[]> {
    methodStub(this, accountId, snapshot);
  }

  public async getHeroById(accountId: string, heroId: number, snapshot?: boolean): Promise<unknown> {
    methodStub(this, accountId, heroId, snapshot);
  }

  public async getAllResources(accountId: string): Promise<unknown> {
    methodStub(this, accountId);
  }

  public async getArena(accountId: string): Promise<unknown> {
    methodStub(this, accountId);
  }

  public async getAcademy(accountId: string): Promise<unknown> {
    methodStub(this, accountId);
  }
}

File: src/Shared/DataModel/APIs/IAccountApi.cs
using System;
using System.Threading.Tasks;

namespace Raid.Toolkit.DataModel
{
    [PublicApi("account-api")]
    public interface IAccountApi
    {
        [PublicApi("updated")]
        event EventHandler<SerializableEventArgs> Updated;

        [PublicApi("getAccountDump")]
        Task<RaidExtractor.Core.AccountDump> GetAccountDump(string accountId);

        [PublicApi("getAllResources")]
        Task<Resources> GetAllResources(string accountId);

        [PublicApi("getAccounts")]
        Task<Account[]> GetAccounts();

        [PublicApi("accountInfo")]
        Task<Account> GetAccount(string accountId);

        [PublicApi("getArtifacts")]
        Task<Artifact[]> GetArtifacts(string accountId);

        [PublicApi("getArtifactById")]
        Task<Artifact> GetArtifactById(string accountId, int artifactId);

        [PublicApi("getHeroes")]
        Task<Hero[]> GetHeroes(string accountId, bool snapshot = false);

        [PublicApi("getHeroById")]
        Task<Hero> GetHeroById(string accountId, int heroId, bool snapshot = false);

        [PublicApi("getArena")]
        Task<ArenaData> GetArena(string accountId);

        [PublicApi("getAcademy")]
        Task<AcademyData> GetAcademy(string accountId);
    }
}
